### PR TITLE
Multi Environment GItlab-CI

### PR DESCRIPTION
**Pull request by arshamparity**
The PR removes the Github workflow for building and pushing docker images as they're going to be built and pushed by the Gitlab-CI from now on:

The Gitlab-CI has three stages:
- **build**
  - builds, tags and pushes Docker images to the Docker repo.
  - When a new tag is pushed (like Github releases), CI tags docker images with the commit tag.
  - When a new commit happens on the Master branch (without any tags!), CI tags docker images with `${CI_COMMIT_SHORT_SHA}-beta`
- **deploy-staging**
  - automatically deploys the images to the staging environment for QA tests.
  - Runs when a new tag is introduced or a commit has happened to the Master branch
- **deploy-production**
  - Runs only on tags
  - The deployment is manual and someone should hit the deploy button to deploy to production

Staging domains:
- kusama-sidecar.parity-stg.parity.io
- polkadot-sidecar.parity-stg.parity.io

Production domains:
- kusama-sidecar.parity-chains.parity.io
- polkadot-sidecar.parity-chains.parity.io

image

### CODE REVIEW

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    name:                 parity-stg
+  rules:
+    - if: '$CI_COMMIT_TAG =~ /^v[0-9]+\.[0-9]+.*$/'       # i.e. v1.0, v2.1rc1
+    - if: '$CI_COMMIT_REF_NAME == "master"'
+
+deploy-production:
+  stage:                  deploy-production
+  <<:                     *deploy-k8s
+  environment:
+    name:                 parity-stg
```
Comment (gabreal) [inline .gitlab-ci.yml:114]: this must be `parity-prod` then

**Review by gabreal — COMMENTED** [review]
lgtm

File: .gitlab-ci.yml (hunk 1 above)
Comment (arshamparity) [reply .gitlab-ci.yml]: Fixed the environment. Thanks

**Review by TarikGul — APPROVED** [review]
lgtm, Thanks for the PR

> The deployment is manual and someone should hit the deploy button to deploy to production

Is this something we should communicate with the CI/CD team for each weeks release?

**Comment by arshamparity** [comment]
> Is this something we should communicate with the CI/CD team for each weeks release?

No. Releasing Docker images will happen automatically so this part is the same as before. But this time instead of using github actions for building those images, we're using gitlab-ci which wouldn't have any differences for developers.

Once the images have been built, they would be deployed to the staging environment automatically without any actions so you can test your code there and make sure it works as expected.

For more safety in production, `deploy-production` stage is configured manually which means that once a new release happens, you should go to the [CI/CD menu in gitlab](https://gitlab.parity.io/parity/substrate-api-sidecar) and hit the play button and it will deploy to production. (you can see the play button in the `deploy-production` stage in the screenshot I've attached)

Remember that `deploy-production` stage would only be available once a new release (actually a new tag) is available while `deploy-staging` stage would happen automatically every time a new release is available or a new commit is pushed to the master branch (regardless of the release).

**Comment by TarikGul** [comment]
> Remember that deploy-production stage would only be available once a new release (actually a new tag) is available while deploy-staging stage would happen automatically every time a new release is available or a new commit is pushed to the master branch (regardless of the release).

Okay sweet this explains it perfectly. I believe we would just need access to the Gitlab enviornment moving forward if we don't already have access. Currently away until the 25th, but I will note this in my backlog to revisit it. Thanks :)

Feel free to merge.

**Comment by arshamparity** [comment]
@paritytech-ci @paritytech/ci may I have a review for this PR?

**Review by gabreal — APPROVED** [review]

**Comment by arshamparity** [comment]
@TarikGul the last note: kusama and polkadot sidecars are connected to a Westend node in the staging for saving resources and costs($).

**Pull request merged**